### ISSUE TITLE
host_callback fails if inside a sharded code block

### ISSUE THREAD

**Issue by gnecula** [google/jax#6538]
If one tries to use host_callback inside a code that is inside `pjit`, they get this error:

```
RuntimeError: Internal: RET_CHECK failure (third_party/tensorflow/compiler/xla/service/spmd/spmd_partitioner.cc:3793) hlo->has_sharding() Side-effect HLO must have sharding: %outfeed.7532 = token[] outfeed(u32[2]{0} %constant.7531, token[] %after-all.641), outfeed_shape=u32[2]{0}, metadata={op_type="outside_call" op_name="xmap(score_batch)/outside_call[ arg_treedef=*\n                                callback=<function _call.<locals>.<lambda> at 0x7fda96a66510>\n                                has_token=True\n                                identity=True ]"
```

**Comment by gnecula** [comment]
I did some digging and the issue is quite complicated, both in terms of implementation and in terms of deciding what semantics we want. One complication is that SPMD sharding requires that the infeed and outfeed have sharding annotations. `lax.infeed` does support some form of sharding annotations, although perhaps not all we need. `lax.outfeed` has no support for sharding annotations at all. 

Let's first figure out what semantics we need, and then see how to implement it. 

For debugging, e.g., `id_print`, one may want to get from each device the slice of the operand that it has, without any additional re-sharding. The is informative, and also cheap because it does not force an `all_gather` on the device. This suggests that we need an `outfeed` without any sharding annotations, but that inherits the sharding of its operand. It is not clear if XLA supports this at the moment. 

The other mode that seems useful is for all devices to cooperate to send a single outfeed to the host, e.g., from device 0, and to wait for a single infeed from the host, on device 0. This would require an `all_gather` before outfeed and a scatter after the infeed. For very large models, this arrangement defeats the partitioning. 

Yes another mode is to put the burden of dealing with different slices entirely on the host code: it will received from all devices their slice of the operand, and will have to send to each device its slice of the result. The host may have to gather all the data and then shard it, but it may not have to do that.

**Comment by gnecula** [comment]
I believe that I fixed this in #6569.
